includes/classes/prod_cat_pulldown.php

```diff
@@ -0,0 +1,186 @@
+				$this->product_vars();
+			}
+
+			$this->run_sql();
+
+			// default styling
+			$this->parameters = '';
+			//$this->parameters = 'required size="15" class="form-control" id="products_id"';
+		}
+
+		private function category_vars()
+		{
+			$this->sql = "SELECT DISTINCT c.categories_id as common_id, c.categories_id, cd.categories_name
+            FROM " . TABLE_CATEGORIES . " c
+            LEFT JOIN " . TABLE_CATEGORIES_DESCRIPTION . " cd ON (c.categories_id = cd.categories_id AND cd.language_id = " . (int)$_SESSION['languages_id'] . ")
+            LEFT JOIN " . TABLE_PRODUCTS_TO_CATEGORIES . " ptoc on (c.categories_id = ptoc.categories_id) "
+				. $this->attributes_join;
+
+			$this->sort = " ORDER BY categories_name";
+			$this->keyword_search_fields = [
+				'cd.categories_name',
+				'c.parent_id',
+				'cd.categories_description',
+				'c.categories_id',
+			];
+		}
+
+		private function product_vars()
+		{
+			$this->keyword_search_fields = [
+				'pd.products_name',
+				'p.products_model',
+				'pd.products_description',
+				'p.products_id',
+			];
+
+			$parm_2 = '%2$s ';
+			if (!$this->show_model) {
+				$parm_2 = '';
+			}
+
+			$this->sort = ' ORDER BY pd.products_name';
+			$this->output_string = '%1$s ' . $parm_2 . '(%3$s)';// format string with name first
+
+			if ($this->order_by == 'model') {
+				$this->sort = ' ORDER by p.products_model';
+				$this->output_string = (!empty($parm_2) ? $parm_2 . '-' : '') . ' %1$s (%3$s)'; // format string with model first
+			}
+
+			$this->sql = "SELECT p.products_id as common_id, pd.products_id, p.products_sort_order, p.products_price, p.products_model, ptc.categories_id
+                FROM " . TABLE_PRODUCTS . " p
+                LEFT JOIN " . TABLE_PRODUCTS_DESCRIPTION . " pd USING (products_id)
+                LEFT JOIN " . TABLE_PRODUCTS_TO_CATEGORIES . " ptc USING (products_id) "
+				. $this->attributes_join .
+				" WHERE pd.language_id = " . (int)$_SESSION['languages_id'];
+		}
+
+		private function run_sql()
+		{
+			global $db, $currencies;
+
+			$search = '';
+			if (!empty($this->keywords)) {
+				$search = zen_build_keyword_where_clause($this->keyword_search_fields,
+					trim(zen_db_input(zen_db_prepare_input($this->keywords))));
+			}
+			$this->sql .= $search;
+
+			switch (true) {
+				case ($this->option_filter === -1): // no selection made: do not list any products
+				case ($this->option_filter === ''): // no selection made: do not list any categories
+					// no selection made yet
+					break;
+				case ((int)$this->option_filter > 0) : // an Option Name was selected: show only products using attributes with this Option Name
+					$this->sql .= $this->condition;
+					$this->sql .= $this->sort;
+					break;
+				default: //legacy: show all products with attributes
+					$this->sql .= $this->sort;
+					break;
+			}
+
+			$results = $db->Execute($this->sql);
+
+			foreach ($results as $result) {
+				if (!in_array($result['common_id'], $this->exclude)) {
```

Comment:
Recommend making this an "early return" characteristic. Instead of grouping everything within this `if`, can test and if met then `continue` the `foreach`:
````
if (in_array($result['common_id'], $this->exclude)) {
    continue;
}
````
And then the remaining code can be left shifted and be outside of this "larger" if group.

Reply:
:+1: